bundler/spec/install/gemfile/force_ruby_platform_spec.rb

```diff
@@ -114,5 +114,48 @@
       expect(the_bundle).to include_gems "depends_on_platform_specific 1.0.0 RUBY"
       expect(the_bundle).to include_gems "platform_specific 1.0.0 #{Bundler.local_platform}"
     end
+
+    it "uses ruby variants for the explicit transitive dependency with a lockile that has only ruby platform" do
+      lockfile <<-L
+        GEM
+          remote: #{file_uri_for(gem_repo4)}
+          specs:
+            dep_level_2 (1.0)
+              depends_on_platform_specific
+            depends_on_platform_specific (1.0)
+              platform_specific
+            platform_specific (1.0)
+
+        PLATFORMS
+          ruby
+
+        DEPENDENCIES
+          dep_level_2
+          platform_specific
+
+        BUNDLED WITH
+            #{Bundler::VERSION}
+      L
+
+      install_gemfile <<-G, :env => { "BUNDLER_SPEC_GEM_REPO" => gem_repo4.to_s }, :artifice => "compact_index", :verbose => true
+        source "#{file_uri_for(gem_repo4)}"
+
+        gem "platform_specific"
+        gem "depends_on_platform_specific"
+      G
+
+      expect(the_bundle).to include_gems "depends_on_platform_specific 1.0.0 #{Bundler.local_platform}"
+      expect(the_bundle).to include_gems "platform_specific 1.0.0 #{Bundler.local_platform}"
```

Comment:
This bug seems to happen when a platform specific variant is pre-installed, the Gemfile is using `:force_ruby_platform => true`, and the lockfile includes "RUBY" platform.

In this case, the locally installed platform specific variant will be selected, and unless we propagate the `force_ruby_platform` flag to it from the Gemfile, it will be happily picked as already installed, and used for materialization.

So, I believe transitive dependencies are unrelated to the fix, and I think the whole spec can be simplified to better illustrate this case like this:

```ruby
it "reinstalls the ruby variant when a platform specific variant is already installed, the lockile has only RUBY platform, and :force_ruby_platform is used in the Gemfile" do
  lockfile <<-L
    GEM
      remote: #{file_uri_for(gem_repo4)}
      specs:
        platform_specific (1.0)

    PLATFORMS
      ruby

    DEPENDENCIES
      platform_specific

    BUNDLED WITH
       #{Bundler::VERSION}
  L

  system_gems "platform_specific-1.0-#{Gem::Platform.local}", :path => default_bundle_path

  install_gemfile <<-G, :env => { "BUNDLER_SPEC_GEM_REPO" => gem_repo4.to_s }, :artifice => "compact_index", :verbose => true
    source "#{file_uri_for(gem_repo4)}"

    gem "platform_specific", :force_ruby_platform => true
  G

  expect(the_bundle).to include_gems "platform_specific 1.0.0 RUBY"
end
```

Reply:
maybe just add that as a second spec, since we also want to test the transitive case, and the behavior could very well end up different with a different implementation

Reply:
Sure, I just don't see it as related to this PR.

I'm also not sure which role `dep_level2` intends to play since it's in the lockfile but not in the gemfile. I guess kit's testing replacing one dependency in the Gemfile with a subdependency of it? If we want to keep it, I'd at least clarify the wording.